cel/env.go

```diff
@@ -205,18 +205,52 @@ func (e *Env) CompileSource(src common.Source) (*Ast, *Issues) {
 }
 
 // Extend the current environment with additional options to produce a new Env.
+//
+// Note, the extended Env value should not share memory with the original. It is possible, however,
+// that a CustomTypeAdapter or CustomTypeProvider options could provide values which are mutable.
+// To ensure separation of state between extended environments either make sure the TypeAdapter and
+// TypeProvider are immutable, or that their underlying implementations are based on the
+// ref.TypeRegistry which provides a Copy method which will be invoked by this method.
 func (e *Env) Extend(opts ...EnvOption) (*Env, error) {
 	if e.chkErr != nil {
 		return nil, e.chkErr
 	}
+	// Copy slices.
+	decsCopy := make([]*exprpb.Decl, len(e.declarations))
+	macsCopy := make([]parser.Macro, len(e.macros))
+	progOptsCopy := make([]ProgramOption, len(e.progOpts))
+	copy(decsCopy, e.declarations)
+	copy(macsCopy, e.macros)
+	copy(progOptsCopy, e.progOpts)
+
+	// Copy the adapter / provider if they appear to be mutable.
+	adapter := e.adapter
+	provider := e.provider
+	adapterReg, isAdapterReg := e.adapter.(ref.TypeRegistry)
+	providerReg, isProviderReg := e.provider.(ref.TypeRegistry)
+	// In most cases the provider and adapter will be a ref.TypeRegistry;
+	// however, in the rare cases where they are not, they are assumed to
+	// be immutable. Since it is possible to set the TypeProvider separately
+	// from the TypeAdapter, the possible configurations which could use a
+	// TypeRegistry as the base implementation are captured below.
+	if isAdapterReg && isProviderReg && adapterReg == providerReg {
+		reg := providerReg.Copy()
+		adapter = reg
+		provider = reg
+	} else if isProviderReg {
```

Comment:
if isAdapterReg && isProviderReg && adapterReg != providerReg should this be making a copy of both providerReg and adapterReg instead of just providerReg?

Reply:
The test here is saying that the adapter and provider are the same `ref.TypeRegistry` so only one copy is needed and desired. 

Thanks for following up!

Reply:
The case I was wondering about is if adapter and provider are both `ref.TypeRegistry` but different ones. I'm not sure if that's an expected case but it looks like we wouldn't copy both of them?

Reply:
Ah, yes. You're right. I have been bouncing between a few different roles here at home and missed that case. It's possible for this to happen, but should be incredibly uncommon. I'll amend the code.